pkg/providers/vsphere/vmlifecycle/create_fastdeploy.go

```diff
@@ -330,45 +331,65 @@ func fastDeployDirect(
 	diskSpecs []*vimtypes.VirtualDeviceConfigSpec,
 	diskFormat vimtypes.DatastoreSectorFormat,
 	dstDiskPaths,
-	srcDiskPaths []string) (*vimtypes.ManagedObjectReference, error) {
+	srcDiskPaths []string,
+	isEncryptedStorageProfile bool) (*vimtypes.ManagedObjectReference, error) {
 
 	logger := pkglog.FromContextOrDefault(ctx).WithName("fastDeployDirect")
 
-	// Copy each disk into the VM directory.
-	if err := fastDeployDirectCopyDisks(
-		ctx,
-		logger,
-		datacenter,
-		configSpec,
-		srcDiskPaths,
-		dstDiskPaths,
-		diskFormat); err != nil {
-
-		return nil, err
-	}
-
-	_, isVMEncrypted := configSpec.Crypto.(*vimtypes.CryptoSpecEncrypt)
-
+	diskCopySpecs := make([]vimtypes.FileBackedVirtualDiskSpec, len(dstDiskPaths))
 	for i := range diskSpecs {
 		ds := diskSpecs[i]
 
 		// Set the file operation to an empty string since the disk already
 		// exists.
 		ds.FileOperation = ""
 
-		if isVMEncrypted {
-			// If the VM is to be encrypted, then the disks need to be updated
-			// so they are not marked as encrypted upon VM creation. This is
-			// because it is not possible to change the encryption state of VM
-			// disks when they are being attached. Instead the disks must be
-			// encrypted after they are attached to the VM.
-			ds.Profile = nil
-			if ds.Backing != nil {
-				ds.Backing.Crypto = nil
+		profile := ds.Profile
+		if len(profile) == 0 {
+			profile = configSpec.VmProfile
+		}
+
+		diskCopySpecs[i] = vimtypes.FileBackedVirtualDiskSpec{
+			VirtualDiskSpec: vimtypes.VirtualDiskSpec{
+				AdapterType: string(vimtypes.VirtualDiskAdapterTypeLsiLogic),
+				DiskType:    string(vimtypes.VirtualDiskTypeThin),
+			},
+			SectorFormat: string(diskFormat),
+			Profile:      profile,
+		}
+
+		// Copy the disks using the same crypto key as the VM if the storage
+		// class is encrypted.
+		if isEncryptedStorageProfile && configSpec.Crypto != nil {
```

Comment:
Looks like `configSpec.Crypto` doesn't have to be not nil, if `ds.Backing.Crypto` is not nil?